### PR TITLE
Added Node name to print() of all Nodes by making to_string() in Object virtual, so it can be overriden in C++.

### LINKED ISSUE #693: Printing node should contain its name
**Describe the project you are working on:**
A game.

**Describe the problem or limitation you are having in your project:**
Sometimes it happens that I have to print a node to see if a proper node is where I want it. My problem is that `print(some_node)` doesn't give really useful information. It gives node type and... instance ID. I don't need the ID, I'd like the name. Most often I can't just do `print(some_node.name)` because `some_node` can be null too (safe navigation operator would help here actually...).

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Doing `print` on node should display its type and name, not type and ID.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
We should either make `to_string` virtual and override it in node or add an `if` checking if the Object is a node. The latter is quite simple to do actually.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
Printing nodes is rather frequent operation IMO.

**Is there a reason why this should be core and not an add-on in the asset library?:**
It's core thing.

### PR DESCRIPTION
**Pull request by EricEzaM**
Closes godotengine/godot-proposals#693
```gdscript
func _ready():
	var special_node = $MySpecialNode
	print(special_node)
	
	var node = Node2D.new()
	print(node)
	
	var node2: Node2D = null
	print(node2)
```
Outputs
```
MySpecialNode [Node2D:23370662652]
[Node2D:23454548333]
Null
```
I did not add position to Node2D/3D/Control as I wasn't sure if it was really need. Some people won't need it, and there is the question of whether to return local/global position (or other info).

### CODE REVIEW

File: scene/main/node.cpp
Comment (akien-mga) [inline scene/main/node.cpp:1960]: The latter part is `Object::to_string()`, so you can use that directly.

Comment (EricEzaM) [reply scene/main/node.cpp]: yeah good point, fixed.

**Comment by ThakeeNathees** [comment]
IMO all string log about the object must be inside the brackets
something like : `[Node2D(MySpecialNode):23370662652]`

**Comment by mhilbrunner** [comment]
The behaviour of all similar things I know of (e.g. Java's `toString()` or C#'s `ToString()`) allow you to *completely* override the result.

E.g. instead of `MySpecialNode [Node2D:23370662652]`, where `MySpecialNode` is the text controlled by the programmer, and `[Node2D:23370662652]` is added regardless, you could completely change the result. And only print `MySpecialNode`. Or build the same message as above manually, if you got access to the Node type name and instance ID.

This is IMO more sensible than "somewhat" allowing a programmer to change the output.

**Comment by EricEzaM** [comment]
I like that idea. I'll make the change when I get up in about 9 hours 😁

**Comment by mhilbrunner** [comment]
All languages that I know of that have this feature let the programmer comoletely override the output. I. e. if I only want it to print `MySpecialNode`, that works. If I still want it to print Node type/name/ID I can do that, too by simply adding them to the string manually (as long as I have access to that data of course).

If we do thos, we should IMO allow that freedom, and not do this halfway and constrain the output.

**Comment by akien-mga** [comment]
@mhilbrunner That's how it works already AFAIK. If you have a script instance, it will return the result from its `to_string()` method, which should be implemented in all script instances to allow users to fully specify the return value.

See `GDScriptInstance::to_string` in `modules/gdscript/gdscript.cpp` for GDScript.

**Comment by mhilbrunner** [comment]
Yes, I was talking about the other comments :) but maybe I misunderstood them.

**Comment by akien-mga** [comment]
Well the discussion is about what the default representation of a Node should be:
https://github.com/godotengine/godot/pull/38819/files#diff-685aaf5e6e44dbc6a5700ebaae623e60R1960

`NodeName [Node:ID]` vs `[Node(NodeName):ID]`.

**Comment by mhilbrunner** [comment]
Then I indeed misunderstood. :) I don't personally care much, both are fine as default for me.

**Comment by EricEzaM** [comment]
Hmm on second thought this morning, I think I will leave it at `NodeName [Node:ID]` unless changes are requested from core maintainers. I think `NodeName [Node:ID]`, it is quicker to find the information you are looking for, whereas in `[Node(NodeName):ID]`, all the information is condensed and I feel that your eyes have to do more work to find what you are after.

Additionally, doing it this way would not allow me to just use the `Object::to_string()` base implementation as akien mentioned in his review. Since the info would need to within square the brackets, I would need to construct the whole thing again. Not a big deal I guess, but still.

**Comment by Xrayez** [comment]
I got surprised that `to_string` is not virtual in the first place. I'd certainly find this useful for C++ modules development, it's strange that it's possible to do this in GDScript and not in C++, unless I'm missing something. 🤔

I'm currently implementing some data structures in a C++ module which I'd like to expose to scripting, having this feature built-in would solve a lot of hacks.

See goostengine/goost#14 and goostengine/goost#12 to justify this change.

**Comment by KoBeWi** [comment]
This probably needs rebase after #42093 was merged.

I'm not sure about the string format. When you print a node as a part of longer string separated by spaces, it's less obvious that `Name` and `[Object:ID]` are related. Maybe they should be linked with `:` or `-`? Or `[Name [Object:ID]]` 🙃

File: scene/main/node.h
Comment (KoBeWi) [inline scene/main/node.h:334]:
```suggestion
	virtual String to_string() override;
```

**Comment by EricEzaM** [comment]
```gdscript
print(self)
# CoolNode:[SpinBox:28454159399]
```

File: scene/main/node.cpp
Comment (akien-mga) [inline scene/main/node.cpp:1834]: Since it's no longer a space the comment is obsolete.

**Review by akien-mga — APPROVED** [review]
There's a comment to amend but otherwise it seems good to me.

**Comment by KoBeWi** [comment]
Is the `get_name() ?` condition even necessary? I don't think we can have an unnamed node, no?

**Comment by akien-mga** [comment]
> Is the `get_name() ?` condition even necessary? I don't think we can have an unnamed node, no?

Node must have a name when in the Scene Tree, but outside it they can be unnamed:
```gdscript
var node = Node.new()
print(node.name)
add_child(node)
print(node.name)
```
Prints:
```

@@2
```

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 3.4.

**Pull request merged**
